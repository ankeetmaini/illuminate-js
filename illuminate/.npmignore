tests/
src/
node_modules/
.babelrc
