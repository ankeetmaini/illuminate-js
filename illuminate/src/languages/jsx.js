import Definition from '../Definition';

import javascript from './javascript';
import markup from './markup';

const jsx = markup.extend(javascript.clone());

jsx.setIn(['tag', 'pattern'], /<\/?[\w\.:-]+\s*(?:\s+(?:[\w\.:-]+(?:=(?:("|')(\\?[\w\W])*?\1|[^\s'">=]+|(\{[\w\W]*?\})))?|\{\.{3}\w+\})\s*)*\/?>/i);
jsx.setIn(['tag', 'inside', 'attr-value', 'pattern'], /=[^\{](?:('|")[\w\W]*?(\1)|[^\s>]+)/i);

jsx.getIn(['tag', 'inside']).insertBefore('attr-name', [
    ['spread', {
        pattern: /\{\.{3}\w+\}/,
        inside: new Definition([
            ['punctuation', /\{|\}|\./],
            ['attr-value', /\w+/]
        ])
    }]
]);

const jsxExpression = jsx.clone();

jsxExpression.delete('punctuation');

jsxExpression.insertBefore('operator', [
    ['punctuation', /=(?={)|[{}[\];(),.:]/]
]);


jsx.getIn(['tag', 'inside']).insertBefore('attr-value', [
    ['script', {
        // Allow for one level of nesting
        pattern: /=(\{(?:\{[^}]*\}|[^}])+\})/i,
        inside: jsxExpression,
        'alias': 'language-javascript'
    }]
]);

export default jsx;
