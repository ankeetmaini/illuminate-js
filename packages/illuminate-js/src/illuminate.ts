import { Token } from './Token';
import { add as addHook, run as runHook, reset as resetHooks } from './hooks';
import { encode } from './utils';

export interface TokenObject {
    pattern: RegExp;
    inside?: Definition;
    lookbehind?: boolean;
    greedy?: boolean;
    alias?: string;
}

export type Tokens = RegExp | TokenObject | Array<RegExp | TokenObject>;

export type Definition = Map<string, Tokens | Map<string, Tokens>>;

export type Plugin = (a: typeof addHook) => void;

export const languages = new Map<string, Definition>();

/**
 * Converts the code to an `Array<string | Token>`, which can be used to write a render.
 * For example, `react-illuminate` uses this to tokenize the code string and render it using react components.
 *
 * @param text    The code to be tokenized
 * @param grammar The definition to be used for tokenizing
 *
 * @private
 *
 * @example
 * import { tokenize } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * const code = `
 * ...
 * `;
 *
 * tokenize(code, javascript); // returns Array<string | Token>
 */
export function tokenize(text: string, grammar: Definition): Array<string | Token> {
    if (!(grammar instanceof Map)) {
        throw new Error('A grammar must be a Map');
    }

    const startArray: Array<string | Token> = [text];

    const rest = grammar.get('rest');

    if (rest instanceof Map) {
        for (const [token, def] of rest as Definition) {
            grammar.set(token, def);
        }

        //TODO: check if we need to shallow copy the grammar
        grammar.delete('rest');
    }

    tokenloop: for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            const pattern = patterns[j];
            let lookbehindLength: number = 0;

            let regex = pattern instanceof RegExp ? pattern : (pattern as TokenObject).pattern;

            const { inside, lookbehind, alias, greedy } = <TokenObject>pattern;

            if (greedy && !regex.global) {
                const flags = (regex.toString().match(/[imuy]*$/) || [''])[0];

                regex = RegExp(regex.source, `${flags}g`);
            }

            for (let i = 0; i < startArray.length; i++) {
                // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                regex.lastIndex = 0;

                let match: RegExpExecArray | string | null = regex.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }

                const stringFrom: number = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const inserts = [];

                if (before) {
                    inserts.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                inserts.push(wrapped);

                if (after) {
                    inserts.push(after);
                }

                startArray.splice(i, 1, ...inserts);
            }
        }
    }

    return startArray;
}

export interface HighlightEnv {
    code: string;
    highlightedCode: string;
    grammar: Definition;
    language: string;
}

/**
 * Highlights the given code. Returns the highlighted code in `string` format.
 *
 * @param text      The code string that is to be highlighted.
 * @param language  Name of the definition to be used. This name corresponds to
 *                  the name given to the definition using `addLanguage`.
 *
 * @example
 * import { addLanguage, highlight } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * const code = `
 * ...
 * `;
 *
 * highlight(code, 'js');
 */
export function highlight(text: string, language: string): string {
    const grammar = languages.get(language);

    if (!grammar) {
        throw new Error(
            `Your are trying to use ${language} language but its definition could not be found. Please make sure to use "addLanguage" to add your definition`
        );
    }

    let env: HighlightEnv = {
        grammar,
        language,
        code: text,
        highlightedCode: ''
    };

    env = <HighlightEnv>runHook('before-highlight', env) || env;

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env = Object.assign({}, env, { highlightedCode: highlightedCode }) || env;
    env = <HighlightEnv>runHook('after-highlight', env) || env;

    return env.highlightedCode; // Specific for plugins
}

/**
 * Helper to add a plugin.
 *
 * @param plugin The plugin to be added.
 *               See [Plugins](./#/plugins/) for more details.
 *
 * @example
 * import { addPlugin } from 'illuminate-js';
 * import { showLanguage } from 'illuminate-js/lib/plugins/showLanguage';
 *
 * addPlugin(showLanguage);
 */
export function addPlugin(plugin: Plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}

/**
 * Removes all the plugins.
 *
 * @example
 * import { resetPlugins } from 'illuminate-js';
 *
 * resetPlugins();
 */
export function resetPlugins() {
    resetHooks();
}

/**
 * Add a language definition to a illuminate.
 * You can add same definition multiple times with different names
 *
 * @param name  The name of the language being added.
 *              This name will be used by `highlight` function.
 * @param def   The language `Definition` to be added.
 *
 * @example
 * import { addLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 */
export function addLanguage(name: string, def: Definition) {
    languages.set(name, def);
}

/**
 * Returns the language definition which was added using `addLanguage`.
 * Returns the `Definition` if found, else returns `undefined`.
 *
 * @param name The name of the language definition
 *
 * @private
 *
 * @example
 * import { addLanguage, getLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * getLanguage('js'); // returns javascript definition
 */
export function getLanguage(name: string): Definition | undefined {
    return languages.get(name);
}
