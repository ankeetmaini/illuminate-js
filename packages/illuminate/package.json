{
  "name": "illuminate-js",
  "version": "0.3.0",
  "description": "Syntax Highlighting for NodeJS",
  "main": "lib/index.js",
  "scripts": {
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vkbansal/illuminate-js.git"
  },
  "author": "Vivek Kumar Bansal <contact@vkbansal.me>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vkbansal/illuminate-js/issues"
  },
  "homepage": "https://github.com/vkbansal/illuminate-js#readme"
}
