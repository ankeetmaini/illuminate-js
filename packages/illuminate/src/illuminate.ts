import { Token } from './Token';
import { add as addHook, run as runHook } from './hooks';
import { encode } from './utils';

export interface TokenObject {
    pattern: RegExp;
    inside?: Map<string, TokenTypes>;
    lookbehind?: boolean;
    greedy?: boolean;
    alias?: string;
}

export type TokenTypes = RegExp | TokenObject | Array<RegExp | TokenObject>;

export type Definition = Map<string, TokenTypes>;

export type Plugin = (a: typeof addHook) => void;

export function tokenize(text: string, grammar: Definition): Array<string | Token> {
    if (!(grammar instanceof Map)) {
        throw new Error('A grammar must be a Map');
    }

    const startArray: Array<string | Token> = [text];

    const rest = grammar.get('rest');

    if (rest instanceof Map) {
        for (const [token, def] of rest as Definition) {
            grammar.set(token, def);
        }

        //TODO: check if we need to shallow copy the grammar
        grammar.delete('rest');
    }

    tokenloop: for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            const pattern = patterns[j];
            let lookbehindLength = 0;

            let regex = pattern instanceof RegExp ? pattern : pattern.pattern;

            const { inside, lookbehind, alias, greedy } = <TokenObject>pattern;

            if (greedy && !regex.global) {
                const flags = (regex.toString().match(/[imuy]*$/) || [''])[0];

                regex = RegExp(regex.source, `${flags}g`);
            }

            for (let i = 0; i < startArray.length; i++) {
                // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                regex.lastIndex = 0;

                let match: RegExpExecArray | string | null = regex.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }

                const stringFrom = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const inserts = [];

                if (before) {
                    inserts.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                inserts.push(wrapped);

                if (after) {
                    inserts.push(after);
                }

                startArray.splice(i, 1, ...inserts);
            }
        }
    }

    return startArray;
}

export interface HighlightEnv {
    code: string;
    highlightedCode: string;
    grammar: Definition;
    language: string;
}

export function highlight(text: string, grammar: Definition, language: string): string {
    const env: HighlightEnv = {
        grammar,
        language,
        code: text,
        highlightedCode: ''
    };

    runHook('before-highlight', env);

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env.highlightedCode = highlightedCode;
    runHook('after-highlight', env);

    return env.highlightedCode; // Specific for plugins
}

export function addPlugin(plugin: Plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}
