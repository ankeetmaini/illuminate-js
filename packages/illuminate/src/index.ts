import {
    TokenObject,
    TokenTypes,
    tokenize,
    Definition,
    Plugin,
    addPlugin,
    highlight
} from './illuminate';

export { TokenObject, TokenTypes, tokenize, Definition, Plugin, addPlugin, highlight };
