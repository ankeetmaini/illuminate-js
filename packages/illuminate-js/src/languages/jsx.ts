import { javascript } from './javascript';
import { markup } from './markup';
import { clone, setIn, getIn, insertBefore } from '../utils';
import { Definition } from '../illuminate';

let jsx: Definition = clone(markup);

javascript.forEach((value, key) => jsx.set(key, value));

setIn(
    jsx,
    ['tag', 'pattern'],
    /<\/?[\w.:-]+\s*(?:\s+(?:[\w.:-]+(?:=(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">=]+|(?:\{\{?[^}]*\}?\})))?|\{\.{3}[a-z_$][\w$]*(?:\.[a-z_$][\w$]*)*\}))*\s*\/?>/i
);
setIn(
    jsx,
    ['tag', 'inside', 'attr-value', 'pattern'],
    /=(?!\{)(?:("|')(?:\\[\s\S]|(?!\1)[^\\])*\1|[^\s'">]+)/i
);

insertBefore(
    getIn<Definition>(jsx, ['tag', 'inside']),
    'attr-name',
    new Map([
        [
            'spread',
            {
                pattern: /\{\.{3}[a-z_$][\w$]*(?:\.[a-z_$][\w$]*)*\}/,
                inside: new Map([['punctuation', /\.{3}|[{}.]/], ['attr-value', /\w+/]])
            }
        ]
    ])
);

let jsxExpression = clone(jsx);
jsxExpression.delete('punctuation');

jsxExpression = insertBefore(
    jsxExpression,
    'operator',
    new Map([['punctuation', /=(?={)|[{}[\];(),.:]/]])
);

insertBefore(
    getIn(jsx, ['tag', 'inside']),
    'attr-value',
    new Map([
        [
            'script',
            {
                // Allow for one level of nesting
                pattern: /=(\{(?:\{[^}]*\}|[^}])+\})/i,
                inside: jsxExpression,
                alias: 'language-javascript'
            }
        ]
    ])
);

export { jsx };
