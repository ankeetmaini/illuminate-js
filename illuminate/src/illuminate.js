import Definition from './Definition';
import Token from './Token';
import { add as addHook, run as runHook } from './hooks';
import { encode } from './utils';
import * as languages from './languages';

export function tokenize(text, grammar) {
    if (!(grammar instanceof Definition)) {
        throw new Error('A grammar must have an instance of Language class');
    }

    const startArray = [text];

    if (grammar.get('rest') instanceof Definition) {
        for (const [token, def] of grammar.get('rest')) {
            grammar.set(token, def);
        }

        grammar.delete('rest');
    }

    tokenloop:
    for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            let pattern = patterns[j];
            let lookbehindLength = 0;
            const { inside, lookbehind, alias, greedy } = pattern;

            if (greedy && !pattern.pattern.global) {
                const flags = pattern.pattern.toString().match(/[imuy]*$/)[0];

                pattern.pattern = RegExp(pattern.pattern.source, `${flags}g`);
            }

            pattern = pattern.pattern || pattern;

            for (let i = 0; i < startArray.length; i++) { // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                pattern.lastIndex = 0;

                let match = pattern.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }


                const stringFrom = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const args = [i, 1];

                if (before) {
                    args.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                args.push(wrapped);

                if (after) {
                    args.push(after);
                }

                startArray.splice(...args);
            }
        }
    }

    return startArray;
}

export function getLanguage(name) {
    if (name in languages) return languages[name];

    return false;
}

export function highlight(text, language) {
    const grammar = getLanguage(language);
    const env = {
        grammar,
        language,
        code: text
    };

    runHook('before-highlight', env);

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env.highlightedCode = highlightedCode;
    runHook('after-highlight', env);

    return env.highlightedCode; // Specific for plugins
}

export function addPlugin(plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}
