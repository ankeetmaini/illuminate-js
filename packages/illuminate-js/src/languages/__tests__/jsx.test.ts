import { jsx } from '../jsx';
import { tokenize } from '../../illuminate';
import { tokenToJson } from './__helpers';

const testCode = `class MyComponent extends Component
    render() {
        const {tabs, ...otherProps} = this.props;
        const {currentTab} = this.state;

        return (
            <TabBar
                {...otherProps}
                currentTab={currentTab}
                onTabClick={this.onTabClick}
                tabs={tabs}
            />
        )
    }
}`;

describe('jsx lang test', () => {
    test('native values', () => {
        expect(tokenToJson(tokenize(testCode, jsx))).toMatchSnapshot();
    });
});
