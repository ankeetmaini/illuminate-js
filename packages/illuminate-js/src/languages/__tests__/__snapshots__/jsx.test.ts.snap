// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx lang test native values 1`] = `
Array [
  Array [
    "keyword",
    "class",
  ],
  Array [
    "class-name",
    Array [
      "MyComponent",
    ],
  ],
  Array [
    "keyword",
    "extends",
  ],
  Array [
    "class-name",
    Array [
      "Component",
    ],
  ],
  Array [
    "function",
    "render",
  ],
  Array [
    "punctuation",
    "(",
  ],
  Array [
    "punctuation",
    ")",
  ],
  Array [
    "punctuation",
    "{",
  ],
  Array [
    "keyword",
    "const",
  ],
  Array [
    "punctuation",
    "{",
  ],
  "tabs",
  Array [
    "punctuation",
    ",",
  ],
  Array [
    "operator",
    "...",
  ],
  "otherProps",
  Array [
    "punctuation",
    "}",
  ],
  Array [
    "operator",
    "=",
  ],
  Array [
    "keyword",
    "this",
  ],
  Array [
    "punctuation",
    ".",
  ],
  "props",
  Array [
    "punctuation",
    ";",
  ],
  Array [
    "keyword",
    "const",
  ],
  Array [
    "punctuation",
    "{",
  ],
  "currentTab",
  Array [
    "punctuation",
    "}",
  ],
  Array [
    "operator",
    "=",
  ],
  Array [
    "keyword",
    "this",
  ],
  Array [
    "punctuation",
    ".",
  ],
  "state",
  Array [
    "punctuation",
    ";",
  ],
  Array [
    "keyword",
    "return",
  ],
  Array [
    "punctuation",
    "(",
  ],
  Array [
    "tag",
    Array [
      Array [
        "tag",
        Array [
          Array [
            "punctuation",
            "<",
          ],
          "TabBar",
        ],
      ],
      Array [
        "spread",
        Array [
          Array [
            "punctuation",
            "{",
          ],
          Array [
            "punctuation",
            "...",
          ],
          Array [
            "attr-value",
            "otherProps",
          ],
          Array [
            "punctuation",
            "}",
          ],
        ],
      ],
      Array [
        "attr-name",
        Array [
          "currentTab",
        ],
      ],
      Array [
        "script",
        Array [
          Array [
            "punctuation",
            "=",
          ],
          Array [
            "punctuation",
            "{",
          ],
          "currentTab",
          Array [
            "punctuation",
            "}",
          ],
        ],
      ],
      Array [
        "attr-name",
        Array [
          "onTabClick",
        ],
      ],
      Array [
        "script",
        Array [
          Array [
            "punctuation",
            "=",
          ],
          Array [
            "punctuation",
            "{",
          ],
          Array [
            "keyword",
            "this",
          ],
          Array [
            "punctuation",
            ".",
          ],
          "onTabClick",
          Array [
            "punctuation",
            "}",
          ],
        ],
      ],
      Array [
        "attr-name",
        Array [
          "tabs",
        ],
      ],
      Array [
        "script",
        Array [
          Array [
            "punctuation",
            "=",
          ],
          Array [
            "punctuation",
            "{",
          ],
          "tabs",
          Array [
            "punctuation",
            "}",
          ],
        ],
      ],
      Array [
        "punctuation",
        "/>",
      ],
    ],
  ],
  Array [
    "punctuation",
    ")",
  ],
  Array [
    "punctuation",
    "}",
  ],
  Array [
    "punctuation",
    "}",
  ],
]
`;
